File: app/src/main/java/com/ch/yoon/kakao/pay/imagesearch/ui/imagedetail/ImageDetailViewModel.java
package com.ch.yoon.kakao.pay.imagesearch.ui.imagedetail;

import android.app.Application;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.ch.yoon.kakao.pay.imagesearch.R;
import com.ch.yoon.kakao.pay.imagesearch.data.model.imagesearch.response.ImageDocument;
import com.ch.yoon.kakao.pay.imagesearch.ui.base.KBaseViewModel;
import com.ch.yoon.kakao.pay.imagesearch.ui.common.livedata.SingleLiveEvent;

/**
 * Creator : ch-yoon
 * Date : 2019-08-05.
 */
public class ImageDetailViewModel extends KBaseViewModel {

    private static final String TAG = ImageDetailViewModel.class.getName();

    @NonNull
    private final MutableLiveData<String> imageUrlLiveData = new MutableLiveData<>();

    @NonNull
    private final SingleLiveEvent<String> docUrlLiveEvent = new SingleLiveEvent<>();
    @NonNull
    private final SingleLiveEvent<String> showMessageLiveEvent = new SingleLiveEvent<>();
    @NonNull
    private final SingleLiveEvent<Void> finishEvent = new SingleLiveEvent<>();

    @Nullable
    private ImageDocument imageDocument;

    public ImageDetailViewModel(@NonNull Application application) {
        super(application);
    }

    @NonNull
    public LiveData<String> observeImageUrl() {
        return imageUrlLiveData;
    }

    @NonNull
    public LiveData<String> observeMoveWebEvent() {
        return docUrlLiveEvent;
    }

    @NonNull
    public LiveData<String> observeShowMessage() {
        return showMessageLiveEvent;
    }

    @NonNull
    public LiveData<Void> observeFinishEvent() {
        return finishEvent;
    }

    public void showImageDetailInfo(@Nullable final ImageDocument imageDocument) {
        if(imageDocument != null) {
            this.imageDocument = imageDocument;
            imageUrlLiveData.setValue(imageDocument.getImageUrl());
        } else {
            handlingUnknownError();
        }
    }

    public void onClickWebButton() {
        if(imageDocument != null) {
            final String docUrl = imageDocument.getDocUrl();
            if(docUrl != null) {
                docUrlLiveEvent.setValue(docUrl);
            } else {
                updateShowMessage(R.string.error_nonexistent_url);
            }
        } else {
            handlingUnknownError();
        }
    }

    public void onClickBackPress() {
        finishEvent.call();
    }

    private void handlingUnknownError() {
        updateShowMessage(R.string.error_unknown_error);
        finishEvent.call();

        Log.d(TAG, "image document is null");
    }

}

File: app/src/main/java/com/ch/yoon/kakao/pay/imagesearch/ui/imagedetail/ImageDetailViewModel.kt
package com.ch.yoon.kakao.pay.imagesearch.ui.imagedetail

import android.app.Application
import android.text.TextUtils
import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.ch.yoon.kakao.pay.imagesearch.R
import com.ch.yoon.kakao.pay.imagesearch.data.repository.model.ImageDocument
import com.ch.yoon.kakao.pay.imagesearch.extention.TAG
import com.ch.yoon.kakao.pay.imagesearch.ui.base.BaseViewModel
import com.ch.yoon.kakao.pay.imagesearch.ui.common.livedata.SingleLiveEvent

/**
 * Creator : ch-yoon
 * Date : 2019-10-29.
 */
class ImageDetailViewModel(application: Application) : BaseViewModel(application) {

    private val _imageUrlInfo = MutableLiveData<String>()
    val imageUrlInfo: LiveData<String> = _imageUrlInfo

    private val _moveToWebEvent = SingleLiveEvent<String>()
    val moveToWebEvent: LiveData<String> = _moveToWebEvent

    private val _finishEvent = SingleLiveEvent<Unit>()
    val finishEvent: LiveData<Unit> = _finishEvent

    private var imageDocument: ImageDocument? = null

    fun showImageDetailInfo(receivedImageDocument: ImageDocument?) {
        receivedImageDocument?.run {
            imageDocument = this
            _imageUrlInfo.value = imageUrl
        } ?: run {
            updateShowMessage(R.string.unknown_error)
            _finishEvent.call()
        }
    }

    fun onClickWebButton() {
        imageDocument?.run {
            if(TextUtils.isEmpty(docUrl)) {
                updateShowMessage(R.string.non_existent_url_error)
            } else {
                _moveToWebEvent.value = docUrl
            }
        }
    }

    fun onClickBackPress() {
        _finishEvent.call()
    }

}